packages/flutter_tools/lib/src/daemon.dart

```diff
@@ -0,0 +1,230 @@
+  /// Stream that contains input to the [DaemonConnection].
+  Stream<Map<String, Object?>> get inputStream;
+
+  /// Outputs a message through the connection.
+  void send(Map<String, Object?> message);
+
+  /// Cleans up any resources used.
+  Future<void> dispose() async { }
+}
+
+/// A [DaemonStream] that uses stdin and stdout as the underlying streams.
+class StdioDaemonStreams extends DaemonStreams {
+  StdioDaemonStreams(): inputStream = _convertInputStream(globals.stdio.stdin);
+
+  @override
+  final Stream<Map<String, Object?>> inputStream;
+
+  @override
+  void send(Map<String, Object?> message) {
+    globals.stdio.stdoutWrite(
+      '[${json.encode(message)}]\n',
+      fallback: (String message, dynamic error, StackTrace stack) {
+        throwToolExit('Failed to write daemon command response to stdout: $error');
+      },
+    );
+  }
+}
+
+/// A [DaemonStream] that uses [Socket] as the underlying stream.
+class TcpDaemonStreams extends DaemonStreams {
+  /// Creates a [DaemonStreams] with an existing [Socket].
+  TcpDaemonStreams(Socket socket) {
+    _socket = Future<Socket>.value(_initializeSocket(socket));
+  }
+
+  /// Connects to a remote host and creates a [DaemonStreams] from the socket.
+  TcpDaemonStreams.connect(String host, int port) {
+    _socket = Socket.connect(host, port).then(_initializeSocket);
+  }
+
+  late final Future<Socket> _socket;
+  final StreamController<Map<String, Object?>> _commands = StreamController<Map<String, Object?>>();
+
+  @override
+  Stream<Map<String, Object?>> get inputStream => _commands.stream;
+
+  @override
+  void send(Map<String, Object?> message) {
+    _socket.then((Socket socket) {
+      try {
+        socket.write('[${json.encode(message)}]\n');
+      } on SocketException catch (error) {
+        globals.logger.printError('Failed to write daemon command response to socket: $error');
+        // Failed to send, close the connection
+        socket.close();
+      }
+    });
+  }
+
+  Socket _initializeSocket(Socket socket) {
+    _commands.addStream(_convertInputStream(socket));
+    return socket;
+  }
+
+  @override
+  Future<void> dispose() async {
+    await (await _socket).close();
+  }
+}
+
+/// Connection between a flutter daemon and a client.
+class DaemonConnection {
+  DaemonConnection({
+    required DaemonStreams daemonStreams,
+    required Logger logger,
+  }): _logger = logger,
+      _daemonStreams = daemonStreams {
+    _commandSubscription = daemonStreams.inputStream.listen(
+      _handleData,
+      onError: (Object error, StackTrace stackTrace) {
+        // We have to listen for on error otherwise the error on the socket
+        // will end up in the Zone error handler.
+        // Do nothing here and let the stream close handlers handle shutting
+        // down the daemon.
+      }
+    );
+  }
+
+  final DaemonStreams _daemonStreams;
+
+  final Logger _logger;
+
+  late final StreamSubscription<Map<String, Object?>> _commandSubscription;
```

Comment:
You can remove this from the initializer so there's not a dangling `late final`.

```suggestion
  late final StreamSubscription<Map<String, Object?>> _commandSubscription = () {
    return _daemonStreams.inputStream.listen(
        _handleData,
        onError: (Object error, StackTrace stackTrace) {
          // We have to listen for on error otherwise the error on the socket
          // will end up in the Zone error handler.
          // Do nothing here and let the stream close handlers handle shutting
          // down the daemon.
        }
    );
  }();
```

Reply:
Unfortunately that doesn't work :(

The `late final` initializer is called when the field is accessed for the first time. In this case it's never accessed until it's about to be disposed.

I still need to initialize it in the constructor.

Reply:
Ah. Sounds good then.